Library/Homebrew/extend/os/linux/diagnostic.rb

```diff
@@ -143,13 +143,24 @@ def check_gcc_dependent_linkage
           return if gcc_dependents.empty?
 
           badly_linked = gcc_dependents.select do |dependent|
-            keg = Keg.new(dependent.prefix)
-            keg.binary_executable_or_library_files.any? do |binary|
-              paths = binary.rpaths
-              versioned_linkage = paths.any? { |path| path.match?(%r{lib/gcc/\d+$}) }
-              unversioned_linkage = paths.any? { |path| path.match?(%r{lib/gcc/current$}) }
-
-              versioned_linkage && !unversioned_linkage
+            # Keg.new() may raise an error if it is not a directory.
+            # As the result `brew doctor` may display `Error: <keg> is not a directory`
+            # instead of proper `doctor` information.
+            # There are other checks that test that, we can skip broken kegs.
+            list_element = dependent # HACK: `brew style` comlains about `begin` otherwise
+            begin
+              keg = Keg.new(dependent.prefix)
+              keg.binary_executable_or_library_files.any? do |binary|
+                paths = binary.rpaths
+                versioned_linkage = paths.any? { |path| path.match?(%r{lib/gcc/\d+$}) }
+                unversioned_linkage = paths.any? { |path| path.match?(%r{lib/gcc/current$}) }
+
+                versioned_linkage && !unversioned_linkage
+              end
+              # It would be nice to add error type to differentiate between
+              # `#{path} is not a valid keg` and `#{path} is not a directory` errors
+            rescue
```

Comment:
A better approach is to make those errors return a `InvalidKegError` or something and catch that.